### PR TITLE
#20 DatabaseExtension upgrade

### PR DESCRIPTION
**Pull request by olenagerasimova**
For #20:
- got rid of static field and method
- upgraded `postgresql` version
- got rid of `flyway` deprecated API usage
- added `itcase` to circle ci

### CODE REVIEW

**Comment by 0crat** [comment]
Job [#51](https://github.com/g4s8/ghman/issues/51) is now in scope, role is `REV`

**Comment by 0crat** [comment]
This pull request [#51](https://github.com/g4s8/ghman/issues/51) is assigned to @victornoel[/z](https://www.0crat.com/u/victornoel), here is [why](https://www.0crat.com/footprint/CJT4R32Q3/2f784d27-33ef-4a47-8398-255b5eea10a8); the budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @g4s8[/z](https://www.0crat.com/u/g4s8) (the architect) right in this ticket; if you decide that this PR should not be accepted ever, also inform the architect; this [blog post](http://www.yegor256.com/2015/02/09/serious-code-reviewer.html) will help you understand what is expected from a code reviewer; there will be no monetary reward for this job

File: src/test/java/com/g4s8/ghman/data/DatabaseExtension.java [hunk 1]
```diff
@@ -73,16 +66,14 @@ public void afterAll(final ExtensionContext context) throws Exception {
...
+        this.postgres.stop();
     }
 
     /**
      * Data source.
      * @return Data source
      */
-    static DataSource dataSource() {
-        return DatabaseExtension.src;
+    DataSource database() {
```
Comment (victornoel) [inline src/test/java/com/g4s8/ghman/data/DatabaseExtension.java]: @olenagerasimova we should make this public

File: src/test/java/com/g4s8/ghman/data/DatabaseExtension.java [hunk 2]
```diff
@@ -18,30 +18,27 @@
...
- * Database abstract case.
+ * JUnit 5 extension to set up database. This extension is meant to be used in
+ * all database related integration tests.
  * @since 1.0
- * @todo #7:30min Find a way to get rid of static field and method of this
- *  class. Note that this class is JUnit 5 Extension and is used to prepare
- *  database to integration tests. See
- *  https://junit.org/junit5/docs/current/user-guide/#extensions-registration-declarative
  */
 final class DatabaseExtension implements BeforeAllCallback, AfterAllCallback {
```
Comment (victornoel) [inline src/test/java/com/g4s8/ghman/data/DatabaseExtension.java]: @olenagerasimova we should make this public

File: src/test/java/com/g4s8/ghman/data/DatabaseExtension.java [hunk 3]
```diff
@@ -51,18 +48,14 @@
...
-        flyway.migrate();
+        this.src.setUrl(this.postgres.start());
+        Flyway.configure().dataSource(this.src).load().migrate();
     }
 
     @Override
     public void afterAll(final ExtensionContext context) throws Exception {
-        try (Connection con = DatabaseExtension.src.getConnection()) {
-            for (final String sql : Arrays.asList(
+        try (Connection con = this.src.getConnection()) {
```
Comment (victornoel) [inline src/test/java/com/g4s8/ghman/data/DatabaseExtension.java]: @olenagerasimova `con` should be final

File: src/test/java/com/g4s8/ghman/data/PgUserITCase.java [hunk 4]
```diff
@@ -29,14 +29,20 @@
  *  methods of {@link com.g4s8.ghman.data.PgUser} class. Do not forget about
  *  all possible exceptions.
  */
-@ExtendWith(DatabaseExtension.class)
 final class PgUserITCase {
 
+    /**
+     * Database.
+     * @checkstyle VisibilityModifierCheck (5 lines)
```
Comment (victornoel) [inline src/test/java/com/g4s8/ghman/data/PgUserITCase.java]: @olenagerasimova why is this needed?

File: src/test/java/com/g4s8/ghman/data/PgUserITCase.java [hunk 5]
```diff
@@ -29,14 +29,20 @@
...
+     */
+    @RegisterExtension
+    static DatabaseExtension source = new DatabaseExtension();
```
Comment (victornoel) [inline src/test/java/com/g4s8/ghman/data/PgUserITCase.java]: @olenagerasimova can we make that non-static? It will simply be instantiated for everytest instead of all the test of the class, it's safer I think.

File: src/test/java/com/g4s8/ghman/data/DatabaseExtension.java (hunk 3 above)
Comment (olenagerasimova) [reply src/test/java/com/g4s8/ghman/data/DatabaseExtension.java]: @victornoel 
`PMD: D:\EO\ghman\src\test\java\com\g4s8\ghman\data\DatabaseExtension.java[57-57]: Unnecessary modifier 'final' on resource specification 'con': resource specifications are implicitly final (UnnecessaryModifier)`

File: .circleci/config.yml [hunk 6]
```diff
@@ -14,5 +14,5 @@ jobs:
           paths:
             - ~/.m2
           key: v1-dependencies-{{ checksum "pom.xml" }}
-      - run: mvn install -Pqulice -Pitcase
+      - run: mvn install -Pqulice
```
Comment (victornoel) [inline .circleci/config.yml]: @olenagerasimova why did you remove it?

File: src/test/java/com/g4s8/ghman/data/PgUserITCase.java (hunk 4 above)
Comment (victornoel) [reply src/test/java/com/g4s8/ghman/data/PgUserITCase.java]: @olenagerasimova you missed this question I think

**Comment by olenagerasimova** [comment]
@victornoel updated, I've made data structure to set up and tear down before/after each test and field not static. Though I do not like this approach:
- integration tests are not supposed to be safe, they are supposed to be as close to reality as possible. In real life we have clean db one time on first application start
- integration tests takes forever (on windows especially)

And no, I cannot make field private, read https://junit.org/junit5/docs/current/user-guide/#extensions-registration-programmatic

File: .circleci/config.yml (hunk 6 above)
Comment (olenagerasimova) [reply .circleci/config.yml]: @victornoel build failed, smth wrong with infrastructure

File: src/test/java/com/g4s8/ghman/data/PgUserITCase.java (hunk 4 above)
Comment (olenagerasimova) [reply src/test/java/com/g4s8/ghman/data/PgUserITCase.java]: @victornoel because checkstyle is complaining and I cannot make this field private

File: .circleci/config.yml (hunk 6 above)
Comment (victornoel) [reply .circleci/config.yml]: @olenagerasimova I've created an issue #52 for this then, thx

File: src/test/java/com/g4s8/ghman/data/PgUserITCase.java (hunk 4 above)
Comment (victornoel) [reply src/test/java/com/g4s8/ghman/data/PgUserITCase.java]: @olenagerasimova I think you can at least make the field final

**Comment by victornoel** [comment]
> Though I do not like this approach:
> 
>     * integration tests are not supposed to be safe, they are supposed to be as close to reality as possible. In real life we have clean db one time on first application start
> 
>     * integration tests takes forever (on windows especially)

@g4s8 WDYT?

**Comment by g4s8** [comment]
@victornoel @olenagerasimova [integration testing](https://en.wikipedia.org/wiki/Integration_testing):
> Integration testing is the phase in software testing in which individual software modules are combined and tested as a group. Integration testing is conducted to evaluate the compliance of a system or component with specified functional requirements.

In other words ITs are not supposed to be as close to reality as possible, ITs should verify that system components are working as expected combined with each other, in our case it's database related objects and Postgres database, so clean table is pretty fine for each IT. Also, each test should be predictable, if we start to reuse tables with data from previous tests, we'll get random false-positive test failures because of test execution order.

What about performance, as I remember, there's some option to tune embedded database on windows, check this: https://github.com/yandex-qatools/postgresql-embedded#known-issues

File: src/test/java/com/g4s8/ghman/data/PgUserITCase.java (hunk 4 above)
Comment (olenagerasimova) [reply src/test/java/com/g4s8/ghman/data/PgUserITCase.java]: @victornoel done

**Comment by victornoel** [comment]
@olenagerasimova thanks
@g4s8 I think it's ok now, please take a look at #52 too that is related to discussions we had here.

**Comment by g4s8** [comment]
@rultor merge

**Comment by 0crat** [comment]
Job was finished in 41 hours, bonus for fast delivery is possible (see [§36](http://www.zerocracy.com/policy.html#36))

**Comment by 0crat** [comment]
@ammaratef45[/z](https://www.0crat.com/u/ammaratef45) please review this job completed by @victornoel[/z](https://www.0crat.com/u/victornoel), as in [§30](http://www.zerocracy.com/policy.html#30); the job will be fully closed and all payments will be made when the quality review is completed

**Comment by 0crat** [comment]
The job [#51](https://github.com/g4s8/ghman/issues/51) is now out of scope

**Comment by ammaratef45** [comment]
@0crat quality good

**Comment by 0crat** [comment]
Payment to `ARC` for a closed pull request, as in [§28](http://www.zerocracy.com/policy.html#28): +10 point(s) just awarded to @g4s8[/z](https://www.0crat.com/u/g4s8)

**Comment by 0crat** [comment]
Quality review completed: +4 point(s) just awarded to @ammaratef45[/z](https://www.0crat.com/u/ammaratef45)

**Comment by 0crat** [comment]
Order was finished, quality is "good": +25 point(s) just awarded to @victornoel[/z](https://www.0crat.com/u/victornoel)
